### PR TITLE
Invalidate key cache if token decoding fails [DP-1008]

### PR DESCRIPTION
**Pull request by skisel-da**
CHANGELOG_BEGIN
CHANGELOG_END

will be merged after #13538 

# Pull Request Checklist

- Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- Include appropriate tests
- Set a descriptive title and thorough description
- Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- Normal production system change, include purpose of change in description
- If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by stefanobaghino-da** [comment]
You don't really need any approval or CI check to merge into a non-`main` branch. 🙂

**Comment by skisel-da** [comment]
yeah, got it ;) thanks

**Comment by stefanobaghino-da** [comment]
But hey, you definitely found a good name for this cache invalidation PR. :trollface:

File: ledger-service/jwt/src/test/scala/com/digitalasset/jwt/JwksSpec.scala [hunk 1]
```diff
@@ -69,6 +74,47 @@ class JwksSpec extends AnyFlatSpec with Matchers {
       s"The token with a mismatching public key should not successfully verify",
     )
   }
+
+  it should "invalidate cached key-id if verification failed" in new JwksSpec.Scope {
```
Comment (mziolekda) [inline ledger-service/jwt/src/test/scala/com/digitalasset/jwt/JwksSpec.scala:78]: No tagging?

Comment (skisel-da) [reply ledger-service/jwt/src/test/scala/com/digitalasset/jwt/JwksSpec.scala:78]: I was not sure what is being exploited in this test in terms of security... do you have an idea?

File: ledger-service/jwt/src/test/scala/com/digitalasset/jwt/JwksSpec.scala [hunk 2]
```diff
@@ -69,6 +74,48 @@ class JwksSpec extends AnyFlatSpec with Matchers {
...
+
+    val updatedJwks = KeyUtils.generateJwks(
+      Map(
+        "test-key-2" -> publicKey2
+      )
+    )
+
+    val updatedServer = SimpleHttpServer.start(updatedJwks, Some(server.getAddress.getPort))
+
+    val expiredToken = generateToken("test-key-1", privateKey1, secondsTillExpiration = -100)
```
Comment (mziolekda) [inline ledger-service/jwt/src/test/scala/com/digitalasset/jwt/JwksSpec.scala:99]: Wouldn't it be better if this failure was done with `test-key-2` instead, to drive the idea home that any misvalidation triggers cache purge?

**Comment by skisel-da** [comment]
As discussed with @mziolekda, this PR opens another security issue that attacker might create invalid tokens which would lead to always invalidating cache, i.e. potential DoS-attack. Closing this PR.

**Pull request closed without merging**